### ISSUE TITLE
Questions regarding class creation

### ISSUE THREAD

**Issue by karussell** [janino-compiler/janino#137]
I found that javaassist reserves space in (PermGen/Metaspace), i.e. for every small class it needs approx 1KB there. But when using janino (via SimpleCompiler.cook) I do not see an increase in Metaspace. Are these classes loaded into the heap instead?

And can I create several different classes with the same class name? It seems this works (`simpleCompiler.getClassLoader().loadClass("com.graphhopper.Test")`) and we seem to be able to create dozens of classes.

Is this safe to do if we call this method basically per-requests (dozens of times per second) or should we prefer a unique name to e.g. avoid confusing the JIT?

**Comment by aunkrig** [comment]
> I found that javaassist reserves space in (PermGen/Metaspace), i.e. for every small class it needs approx 1KB there. But when using janino (via SimpleCompiler.cook) I do not see an increase in Metaspace. Are these classes loaded into the heap instead?

I didn't know that you can "reserve space in PermGen/Metaspace" from Java code!? Please enlighten me. All that Janino does is invoke `ClassLoader.defineClass(byte[])` (in `ByteArrayClassLoader.findClass()`), and Bob's your uncle.

My guess is that the PermGen/Metaspace consumption is only a side effect of what javassist does.

> And can I create several different classes with the same class name

Java enforces that there can be only one class with a given name per class loader instance. However, each instance of `SimpleCompiler`, `ClassBodyEvaluator`, `ScriptEvaluator` and `ExpressionEvaluator` creates its own class loader instance, and defines the classes it generates in that class loader. Thus, classes with equal name can exist in different SCs, CBEs, SEs and EEs, but not within *one*.

> Is this safe to do if we call this method basically per-requests (dozens of times per second) or should we prefer a unique name to e.g. avoid confusing the JIT?

I don't quite understand what you mean. `simpleCompiler.getClassLoader()` will always return the same class loader, and `...loadClass()` will always return the same `java.lang.Class` for equal class name.

**Comment by karussell** [comment]
> I didn't know that you can "reserve space in PermGen/Metaspace" from Java code!? Please enlighten me.

I have no idea about this topic. So all I did was comparing the Janino class creation with Java assist and Janino does not seem to use Metaspace.

> My guess is that the PermGen/Metaspace consumption is only a side effect of what javassist does.

Ok :) I just thought that as soon as a class is loaded some memory in Metaspace is required. See the [hotspot wiki](https://wiki.openjdk.java.net/display/HotSpot/Metaspace) where it says: "Class metadata are allocated when classes are loaded." ... ah but maybe because of throwing away the SimpleCompiler the following kicks in "Their lifetime is usually scoped to that of the loading classloader"

> creates its own class loader instance, and defines the classes it generates in that class loader.

Ok, so doing `new SimpleCompiler().cook` allows us to use the same class name. Are there advantages or disadvantages of this compared to reusing the same SimpleCompiler instance?

> I don't quite understand what you mean

I now think this is unrelated. We stumbled over some [weird JVM behaviour](https://bugs.java.com/bugdatabase/view_bug.do?bug_id=JDK-8247618) before, that led to a "deoptimization" of hot code (shared between different subclasses). And so the creation of a second subclass created a small slow down in the first subclass.
